crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -206,13 +206,51 @@ impl<'db> SemanticIndexBuilder<'db> {
     }
 
     fn pop_scope(&mut self) -> FileScopeId {
-        let ScopeInfo { file_scope_id, .. } =
-            self.scope_stack.pop().expect("Root scope to be present");
-        let children_end = self.scopes.next_index();
-        let scope = &mut self.scopes[file_scope_id];
-        scope.descendents = scope.descendents.start..children_end;
         self.try_node_context_stack_manager.exit_scope();
-        file_scope_id
+
+        let ScopeInfo {
+            file_scope_id: popped_scope_id,
+            ..
+        } = self
+            .scope_stack
+            .pop()
+            .expect("Root scope should be present");
+
+        let mut scopes = std::mem::take(&mut self.scopes);
+        let children_end = scopes.next_index();
+        let popped_scope = &mut scopes[popped_scope_id];
+        popped_scope.extend_descendents(children_end);
+
+        // If the scope that we just popped off is an eager scope, we need to "lock" our view of
+        // which bindings and/or declarations reach each of the uses in the scope.
+        if let Some(eager_nested_scope) = popped_scope.node().as_eager_nested_scope() {
+            for nested_symbol in self.symbol_tables[popped_scope_id].symbols() {
+                // Loop through each enclosing scope, looking for the innermost one that binds or
+                // declares the symbol.
+                for enclosing_scope_info in self.scope_stack.iter().rev() {
+                    let enclosing_scope_id = enclosing_scope_info.file_scope_id;
+                    let enclosing_symbol_table = &self.symbol_tables[enclosing_scope_id];
+                    if let Some(enclosing_symbol_id) =
+                        enclosing_symbol_table.symbol_id_by_name(nested_symbol.name())
+                    {
+                        let enclosing_symbol = enclosing_symbol_table.symbol(enclosing_symbol_id);
+                        if enclosing_symbol.is_bound() || enclosing_symbol.is_declared() {
```

Comment:
I'm not sure about the `is_declared()` here. It looks to me like we only snapshot bindings, not declarations, and the corresponding logic in `TypeInferenceBuilder::infer_name_load` only checks `is_bound()`, so it seems like including `is_declared` is a discrepancy here.

It seems to me that in general we want to treat an eager nested scope as if it were a reference in that enclosing scope, not a "public" reference, which would suggest that only considering bindings is correct, and we shouldn't consult `is_declared()` here. But I think we should have a test where a nonlocal eager reference goes "through" a scope where a name is declared but not bound, to confirm the desired behavior here.

Reply:
I've removed the `is_declared` part, but haven't been able to construct a test case for it yet.  I'm not sure how to construct a declaration-that-isn't-a-binding in an eager scope.  (Class definitions are eager scopes, but a declaration in the class body is explicitly not visible to nested scopes.  Comprehensions are eager scopes, but those contain an _expression_, not a suite of statements, and there isn't a declaration expression.)